python/ray/llm/_internal/batch/stages/sglang_engine_stage.py

```diff
@@ -0,0 +1,367 @@
+"""The stage that runs SGLang engine."""
+
+import asyncio
+import logging
+import time
+import uuid
+from enum import Enum
+from pydantic import BaseModel, root_validator
+from typing import Any, Dict, AsyncIterator, Optional, List, Tuple, Type
+
+from ray.llm._internal.utils import try_import
+from ray.llm._internal.batch.stages.base import (
+    StatefulStage,
+    StatefulStageUDF,
+)
+
+sgl = try_import("sglang")
+
+logger = logging.getLogger(__name__)
+
+
+class SGLangTaskType(str, Enum):
+    """The type of task to run on the SGLang engine."""
+
+    """Generate text."""
+    GENERATE = "generate"
+
+    # TODO: Add support for the following tasks.
+    """Generate embeddings."""
+    EMBED = "embed"
+
+    """Classify the quality of pairwise generations."""
+    REWARD = "reward"
+
+
+class SGLangEngineRequest(BaseModel):
+    """A request to the SGLang engine."""
+
+    # The request ID for the LLM engine (unique per replica).
+    request_id: int
+    # The index of the request in the batch.
+    idx_in_batch: int
+    # The input prompt.
+    text: Optional[str]
+    # Alternative to text. Specify the input as token IDs instead of text.
+    input_ids: Optional[List[int]]
+    # The sampling parameters (more details can be seen in https://docs.sglang.ai/backend/sampling_params.html).
+    sampling_params: Optional[Dict[str, Any]]
```

Comment:
Is it possible to use the same names as the ones in `vllm_engine_stage` so that migration / inheritance in the future could work out of the box?

Reply:
input_ids --> prompt_token_ids
sampling_params --> params

Reply:
I'm actually debating whether we should share as much code as possible between vLLM and SGLang stages, or we should intentionally duplicate everything to make sure they can be improved in a flexible way. After all, these names are not standard and could be changed over time.

Reply:
I think we should duplicate to keep the flexibility but keep the names consistent as much as possible to make it easier for someone to compare in case there are comparison points.

Reply:
Done